lib/bootstrapper.js

```diff
@@ -0,0 +1,210 @@
+ * @param options
+ * @constructor
+ */
+function Bootstrapper(options) {
+  this.plugins = [];
+  options = options || {};
+
+  if (typeof options === 'string') {
+    options = { appRootDir: options };
+  }
+
+  var appRootDir = options.appRootDir = options.appRootDir || process.cwd();
+  var env = options.env || process.env.NODE_ENV || 'development';
+
+  var appConfigRootDir = options.appConfigRootDir || appRootDir;
+
+  options.rootDir = appConfigRootDir;
+  options.env = env;
+  this.options = options;
+
+  this.phases = options.phases || builtinPhases;
+  this.builtinPlugins = options.plugins || builtinPlugins;
+  assert(Array.isArray(this.phases), 'Invalid phases: ' + this.phases);
+  assert(Array.isArray(this.plugins), 'Invalid plugins: ' +
+    this.builtinPlugins);
+
+  var self = this;
+  self.builtinPlugins.forEach(function(p) {
+    var factory = require('./plugins/' + p);
+    self.use('/boot/' + p, factory(options));
+  });
+
+  try {
+    loadAndRegisterPlugins(self, options);
+  } catch (err) {
+    debug(err);
+  }
+}
+
+/**
+ * Register a handler to a given path
+ * @param {String} path
+ * @param {Function} handler
+ */
+Bootstrapper.prototype.use = function(path, handler) {
+  var plugin = {
+    path: path,
+    handler: handler,
+  };
+  this.plugins.push(plugin);
+};
+
+/**
+ * Get a list of plugins for the given path
+ * @param {String} path
+ * @returns {*}
+ */
+Bootstrapper.prototype.getPlugins = function(path) {
+  if (path[path.length - 1] !== '/') {
+    path = path + '/';
+  }
+  return this.plugins.filter(function(p) {
+    return p.path.indexOf(path) === 0;
+  });
+};
+
+/**
+ * Get a list of extensions for the given path
+ * @param {String} path
+ * @returns {*}
+ */
+Bootstrapper.prototype.getExtensions = function(path) {
+  if (path[path.length - 1] !== '/') {
+    path = path + '/';
+  }
+  return this.plugins.filter(function(p) {
+    if (p.path.indexOf(path) === -1) return false;
+    var name = p.path.substring(path.length);
+    return name && name.indexOf('/') === -1;
+  });
+};
+
+/**
+ * Add more phases. The order of phases is decided by the sequence of phase
+ * names
+ * @param {String[]} phases An array of phase names
+ * @returns {String[]} New list of phases
+ */
+Bootstrapper.prototype.addPhases = function(phases) {
+  this.phases = utils.mergePhaseNameLists(this.phases, phases || []);
+  return this.phases;
+};
+
+function pluginIteratorFactory(context, phase) {
+  return function(plugin, done) {
+    var result;
+    if (typeof plugin.handler[phase] !== 'function') {
+      debug('Skipping %s.%s', plugin.handler.name, phase);
+      return done();
+    }
+    debug('Invoking %s.%s', plugin.handler.name, phase);
+    try {
+      if (plugin.handler[phase].length === 2) {
+        plugin.handler[phase](context, done);
+      } else {
+        result = plugin.handler[phase](context);
+        if (typeof Promise !== 'undefined') {
+          Promise.resolve(result).then(function(value) {
```

Comment:
Is this trying to implement support for handler functions returning a promise? Here is the version that we use elsewhere in LoopBack ([example](https://github.com/strongloop/loopback-datasource-juggler/blob/85625829e065bd2a2a9fbbe9ccf47edb75e4d063/lib/observer.js#L104-L109), it is addressing many subtle edge cases that are handled incorrectly by this code.

```js
if (result && typeof result.then === 'function') {
  result.then(
    function onPluginPhaseResolved(value) {
      done(null, value); 
    },
    function onPluginPhaseRejected(err) {
      debug('Unable to invoke %s.%s()', plugin.name, phase, err); 
      done(err); 
    });
} else {
  done(null, result);
}
```